switch to per-tenant attach/detach

LizardWizzard opened the pull request:
download operations of all timelines for one tenant are now grouped
together so when attach is invoked pageserver downloads all of them
and registers them in a single apply_sync_status_update call so
branches can be used safely with attach/detach

SomeoneToIgnore commented on pageserver/src/storage_sync/download.rs:
```diff
@@ -23,6 +27,105 @@ use super::{
+}
+
 /// Retrieves index data from the remote storage for a given timeline.
 pub async fn download_index_part<P, S>(
```
That is not used anywhere now, so `pub` can be removed.

SomeoneToIgnore commented on pageserver/src/storage_sync/download.rs:
```diff
@@ -23,6 +27,105 @@ use super::{
+    index_parts
+}
+
+pub async fn download_tenant_index_parts<P, S>(
```
I think we should name it differently (or its `try_` counterpart) to emphasize that in this case, we actually do S3 list and download all timelines, while in the other one only the given ids.

This method is rather dangerous and expensive from S3 access point of view, compared to the other one, so also the `routes::try_download_tenant_index` function is worth renaming to emphasize that. 
We could even more the whole logic there, as dangerous and not intended to be used in pageserver regularly.

LizardWizzard replied:
Renamed, added a bunch of comments

SomeoneToIgnore commented on pageserver/src/storage_sync/download.rs:
```diff
@@ -23,6 +27,105 @@ use super::{
+        )
+    })?;
+    let timelines = storage
+        .list_prefixes(Some(tenant_storage_path))
```
`let timelines = storage.list_prefixes(..).await..?` 

IMO, it's neither timelines (rather, timeline_directory_keys?) and nor prefixes, so `list_prefixes` could also be named somehow more accurate.

SomeoneToIgnore commented on pageserver/src/storage_sync.rs:
```diff
@@ -918,16 +920,59 @@ fn storage_sync_loop<P, S>(
                     );
-                    // Batch timeline download registration to ensure that the external registration code won't block any running tasks before.
-                    apply_timeline_sync_status_updates(conf, &index, new_timeline_states);
+                    let index_accessor = runtime.block_on(index.write());
```
I also see now that the old `apply_timeline_sync_status_updates` collected all tenant updates into one batch, if that could be returned, we would need to do only one blocking call after collecting the data needed, naturally solving the issue.

LizardWizzard replied:
Done

SomeoneToIgnore commented on pageserver/src/walreceiver.rs:
```diff
@@ -182,7 +182,7 @@ async fn wal_receiver_main_thread_loop_step<'a>(
             match update {
                 // Timeline got detached, stop all related tasks and remove public timeline data.
-                LocalTimelineUpdate::Detach(id) => {
+                LocalTimelineUpdate::Detach(id, join_sender) => {
```
It's a bit hard to understand what is this sender for before coming here, so a named enum field might be a bit more helpful.

If I understand it right, it's used once only to ensure that the drop happened, so we can presumably use a `oneshot` flavor from the stdlib?

LizardWizzard replied:
AFAIK there is no separate oneshot channel in std. You probably confused it with the tokio one. The default synchronous mpsc from std is oneshot by default and is transparently upgraded: https://doc.rust-lang.org/src/std/sync/mpsc/mod.rs.html#713-716

LizardWizzard replied:
But I agree, naming can be improved

LizardWizzard replied:
Changed it to be more verbose

SomeoneToIgnore reviewed (commented):
Looks good.

I think some extra love could be added to the index loading, but otherwise a good first iteration.

LizardWizzard commented on libs/remote_storage/src/s3_bucket.rs:
```diff
@@ -250,6 +268,77 @@ impl RemoteStorage for S3Bucket {
+        &self,
+        prefix: Option<Self::RemoteObjectId>,
+    ) -> anyhow::Result<Vec<Self::RemoteObjectId>> {
+        let list_prefix = match prefix {
```
Have few thoughts on this: 
* Join is indeed a bit cleaner, but it will replace only part of the logic. Trailing separator logic will remain in this function
* Regarding `impl RemoteObjectId for PathBuf` here join will clash with the PathBuf::join whci his problematic

Any ideas to make it look better?

SomeoneToIgnore replied:
Maybe it was a too idealistic thought, since indeed the first bullet would require us to distinguish between files and directories, that could take a bit more efforts to implement.
Yet I think it's the `join`'s responsibility to ensure all separators are in place, so not fully agree with that first bullet.

As for the second bullet, it could be just renamed into something else, don't see much issues with that.

But it starts to look like a big complication just for the readability, so there might be a better way to do things here.

LizardWizzard replied:
I mean the last separator. To me there is no sense in join being responsible for path to end with a separator or not. Because this actually changes the API response. So for some cases it may be desired to end with a separator, for some not and IMO join is not the right place to make this decision

LizardWizzard commented on pageserver/src/storage_sync.rs:
```diff
@@ -918,16 +920,59 @@ fn storage_sync_loop<P, S>(
+                                .keys()
+                                .copied()
+                                .map(|timeline_id| {
+                                    (timeline_id, TimelineSyncStatusUpdate::Downloaded)
```
Done

LizardWizzard commented on pageserver/src/http/openapi_spec.yml:
```diff
@@ -186,12 +185,27 @@ paths:
+          type: string
+          format: hex
+    post:
+      description: Deprecated
```
Fixed

LizardWizzard commented on libs/remote_storage/src/s3_bucket.rs:
```diff
@@ -117,6 +117,24 @@ impl S3ObjectKey {
+    /// Turn a/b/c or a/b/c/ into c
+    fn object_name(&self) -> Option<&str> {
+        // corner case
+        if &self.0 == "/" {
```
Done in a bit weird way, hope for char::to_string to become const at some point :)

LizardWizzard commented:
@SomeoneToIgnore I've adjusted everything based on your suggestions, one discussion is still open. We still need timeline/delete so I brought back timeline/detach and renamed it to timeline/delete with some adjustments. Currently it wont delete data from s3, but I think it should. I think this can be added separately. I'm focusing on hardening and follow ups. Better guarantees for moving branches etc

LizardWizzard commented on pageserver/src/storage_sync/download.rs:
```diff
@@ -23,8 +27,110 @@ use super::{
+
+    for timeline_remote_storage_key in timelines {
+        let object_name = timeline_remote_storage_key.object_name().ok_or_else(|| {
+            anyhow::anyhow!("failed to get timeline id for remote tenant {tenant_id}")
```
Hmm, it's hard to guess what the situation might be, having extra files in tenant dir on s3 is not something usual to me, so IMO error is appropriate here, because otherwise we wont notice that something weird is going on. So I would rather take more defensive approach, leaving an error and decide whats better in the concrete case if it happens. But I'm not insisting.

SomeoneToIgnore replied:
Neither do I, so we can leave it as is.
